File: services/data-processor/internal/models/market_data.go
package models

import (
    "time"
)

// MarketData for kafka consumer (no GORM)
type MarketData struct {
    ID            int64     `json:"id"`
    Symbol        string    `json:"symbol"`
    Price         float64   `json:"price"`
    Volume        *int64    `json:"volume"`
    ChangeAmount  *float64  `json:"changeamount"`
    ChangePercent *float64  `json:"changepercent"`
    Timestamp     time.Time `json:"timestamp"`
    DataType      string    `json:"datatype"`
}

// MarketDataGorm GORM
type MarketDataGorm struct {
    ID            int64      `gorm:"primaryKey;autoIncrement"`
    Symbol        string     `gorm:"type:varchar(20);not null;index"`
    Price         float64    `gorm:"type:decimal(12,4);not null"`
    Volume        *int64     `gorm:"type:bigint"`
    ChangeAmount  *float64   `gorm:"type:decimal(10,4);column:changeamount"`
    ChangePercent *float64   `gorm:"type:decimal(8,4);column:changepercent"`
    Timestamp     time.Time  `gorm:"type:timestamp with time zone;index;default:CURRENT_TIMESTAMP"`
    DataType      string     `gorm:"type:varchar(10);default:'stock'"`
    CreatedAt     time.Time  `gorm:"autoCreateTime"`
    UpdatedAt     time.Time  `gorm:"autoUpdateTime"`
}

func (MarketDataGorm) TableName() string {
    return "market_data"
}

// ToGorm parse MarketData to MarketDataGorm
func (m *MarketData) ToGorm() *MarketDataGorm {
    return &MarketDataGorm{
        ID:            m.ID,
        Symbol:        m.Symbol,
        Price:         m.Price,
        Volume:        m.Volume,
        ChangeAmount:  m.ChangeAmount,
        ChangePercent: m.ChangePercent,
        Timestamp:     m.Timestamp,
        DataType:      m.DataType,
    }
}

// FromGorm parse MarketDataGorm to MarketData
func (m *MarketData) FromGorm(gormData *MarketDataGorm) {
    m.ID = gormData.ID
    m.Symbol = gormData.Symbol
    m.Price = gormData.Price
    m.Volume = gormData.Volume
    m.ChangeAmount = gormData.ChangeAmount
    m.ChangePercent = gormData.ChangePercent
    m.Timestamp = gormData.Timestamp
    m.DataType = gormData.DataType
}

File: services/data-processor/src/entities/MarketData.ts
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, UpdateDateColumn, Index } from 'typeorm';

@Entity('market_data')
@Index(['symbol'])
@Index(['timestamp'])
export class MarketData {
  @PrimaryGeneratedColumn('increment')
  id!: number

  @Column({ type: 'varchar', length: 20 })
  symbol!: string

  @Column({ type: 'decimal', precision: 12, scale: 4 })
  price!: number

  @Column({ type: 'bigint', nullable: true })
  volume?: number

  @Column({ name: 'changeamount', type: 'decimal', precision: 10, scale: 4, nullable: true })
  changeAmount?: number

  @Column({ name: 'changepercent', type: 'decimal', precision: 8, scale: 4, nullable: true })
  changePercent?: number

  @Column({ type: 'timestamp with time zone' })
  timestamp!: Date

  @Column({ name: 'datatype', type: 'varchar', length: 10, default: 'stock' })
  dataType!: string

  @CreateDateColumn({ name: 'created_at', type: 'timestamp with time zone' })
  createdAt!: Date

  @UpdateDateColumn({ name: 'updated_at', type: 'timestamp with time zone' })
  updatedAt!: Date
}

export interface CreateMarketDataDto {
  symbol: string
  price: number
  volume?: number
  changeAmount?: number
  changePercent?: number
  timestamp: Date
  dataType?: string
}